### ISSUE TITLE
How to convert trained CycleGAN model to ONNX, eventually coreML model

### ISSUE THREAD

**Issue by HeoJinLareine** [junyanz/pytorch-CycleGAN-and-pix2pix#852]
Thank's to you, I've achieved my goal through your cycleGAN model.

I hope to convert cycleGAN model to ONNX model, eventually coreML model.

On your code, which file should I try convert?

And any advice how to do it?

Once again, I'm really thanks to you!

**Comment by CesarCoelho** [comment]
Hi HeoJinLareine,

We tried to convert the CycleGAN model to ONNX but we got some problems:

First:
TestModel object has no attribute ‘state_dict’

After changing the net into this attribute, we got:
“forward() got an unexpected keyword argument: ‘keep_vars’”
This seems to be a generic function call to neural networks, so it is a bit fuzzy to me which of the actual implementations is really called there (I would assume TestModel). However, I added this argument to all of those function definitions, but still get this error. 

Did you try to do this? If so, can you share your solution?

Best regards,
César Coelho

**Comment by rtolps** [comment]
Hi! I'm seeking to convert this model to ONNX too. I wouldn't even know where to start though.

**Comment by Lia-C** [comment]
I as well!

**Comment by Starkguava** [comment]
+1

**Comment by tongchangD** [comment]
+1

**Comment by youjinChung** [comment]
+1

**Comment by tongchangD** [comment]
> +1

You can fix this error by compiling and installing PyTorch yourself

**Comment by PL23K** [comment]
Here is my code.
Export:
`
import os
import torch
from options.test_options import TestOptions
from data import create_dataset
from models import create_model
from util.visualizer import save_images
from util import html


if __name__ == '__main__':
    opt = TestOptions().parse()  # get test options
    # hard-code some parameters for test
    opt.num_threads = 0   # test code only supports num_threads = 0
    opt.batch_size = 1    # test code only supports batch_size = 1
    opt.serial_batches = True  # disable data shuffling; comment this line if results on randomly chosen images are needed.
    opt.no_flip = True    # no flip; comment this line if results on flipped images are needed.
    opt.display_id = -1   # no visdom display; the test code saves the results to a HTML file.
    dataset = create_dataset(opt)  # create a dataset given opt.dataset_mode and other options
    model = create_model(opt)      # create a model given opt.model and other options
    model.setup(opt)               # regular setup: load and print networks; create schedulers
    
    model.eval()
    
    print(model.netG_A) # netG for pix2pix; netG_A for cyclegan
    dummy = torch.randn(1, 3, 256, 256)
    torch.onnx.export(model.netG_A, dummy, './out.onnx')
`

Usage c++ :
`cv::Mat HbGan::Gan(const cv::Mat& image , int size) {
    // Resize and preprocess the image
    cv::Mat image_resized, rgb_image;
    cv::resize(image, image_resized, cv::Size(size, size));
    image_resized.convertTo(rgb_image, CV_32F, 1.0 / 255);
    // Create input tensor
    std::array<int64_t, 4> input_shape = {1, 3, size, size};
    size_t input_tensor_size = 1 * 3 * size * size;
    std::size_t counter = 0;
    std::vector<float> input_tensor_values(input_tensor_size);
    for (int c = 0; c < rgb_image.channels(); ++c) {
        for (int h = 0; h < rgb_image.rows; ++h) {
            for (int w = 0; w < rgb_image.cols; ++w) {
                input_tensor_values[counter] = rgb_image.at<cv::Vec3f>(h, w)[c];
                counter++;
            }
        }
    }

    Ort::Value input_tensor = Ort::Value::CreateTensor<float>(memory_info, input_tensor_values.data(), input_tensor_size, input_shape.data(), input_shape.size());
    // Run the model
    auto output_tensors = session.Run(Ort::RunOptions{ nullptr }, input_node_names.data(), &input_tensor, 1, output_node_names.data(), 1);

    // Process the output
    float* output_data = output_tensors.front().GetTensorMutableData<float>();
    // 其它代码
    cv::Mat data_out(size, size, CV_32F), out_image;  // CV_32F maybe error, u can try another. I was written for grayscale images.
    memcpy(data_out.data, output_data, size*size*3*sizeof(float));
    data_out.convertTo(out_image, CV_8UC3, 255);
    return out_image;
}`
